### PR TITLE
Wayland backend xdg-surface and wrl-layer

### PR DESCRIPTION
**Pull request by nongio**

### CODE REVIEW

File: Source/cairo/WaylandCairoSurface.m
Comment (fredkiefer) [inline Source/cairo/WaylandCairoSurface.m:211]: Do we need the attach call here even when it was already called in -initWithDevice:?

I would expect that we have a flag that tells us, whether this was already called if there is no consistent order in which these methods get called.

File: Source/wayland/WaylandServer+Cursor.m
Comment (fredkiefer) [inline Source/wayland/WaylandServer+Cursor.m:311]: We also support the version of the event without the button number. Just use that if you prefer. The more important bit is to set the event type correctly as for the first three buttons we already have the information there.

**Comment by fredkiefer** [comment]
This is way too much code for me to read through. As Wayland support is still experimental in GNUstep I would say we just take your code as it seems to work better than the previous one.
It would be great if you could clean up the code a bit. That is remove commented out portions and try to stick to the GNUstep coding standard. (See http://www.gnustep.org/resources/documentation/Developer/CodingStandards/coding-standards.ps) Yes, this is horrible and really old. But having some standard saves us from discussing this every year.

**Comment by nongio** [comment]
thanks for checking it @fredkiefer I updated the code using clang-format with this configuration file:
https://github.com/gnustep/libs-base/blob/master/.clang-format
let me know if it can go like this

File: Headers/wayland/WaylandServer.h
Comment (fredkiefer) [inline Headers/wayland/WaylandServer.h:8]: You should leave the old dates in, to show when the contribution had happened.

File: Source/cairo/WaylandCairoSurface.m
Comment (fredkiefer) [inline Source/cairo/WaylandCairoSurface.m:48]: Is this a typo or intentional? Most likely it should be "gnustep"

**Review by fredkiefer — APPROVED** [review]
The formatting looks a lot better now.
It would be great if you could re-add the original dates and maybe fix the typo. After that merging is fine for me.

File: Headers/wayland/WaylandServer.h
Comment (nongio) [reply Headers/wayland/WaylandServer.h:8]: 👍

File: Source/cairo/WaylandCairoSurface.m
Comment (nongio) [reply Source/cairo/WaylandCairoSurface.m:48]: :) indeed, I am going to amend the previous commit

File: Source/cairo/WaylandCairoSurface.m
Comment (nongio) [reply Source/cairo/WaylandCairoSurface.m]: it might be superfluous this code, i need to debug it and add a comment for this edge case, the flag used elsewhere is `window->buffer_needs_attach`

**Comment by ivucica** [comment]
nit: That should likely be November 2021, not November 2016.

sent from phone

File: Source/cairo/WaylandCairoSurface.m
Comment (nongio) [reply Source/cairo/WaylandCairoSurface.m]: @fredkiefer i added a comment in the code to explain better when this is needed:
usually the attach() is handled in the configure callback, when resizing a window already configured then the new buffer created should be attached immediately. maybe we could move the attach outside of the surface class...

File: Source/wayland/WaylandServer+Cursor.m
Comment (nongio) [reply Source/wayland/WaylandServer+Cursor.m]: I think that the current code structure provides a good separation of concerns: Cursors, Keyboard, Layer/Xdg protocols, Seat and Output. Starting from this we could start collecting feedback on the separate areas and tackle them 1 by 1

Comment (fredkiefer) [reply Source/wayland/WaylandServer+Cursor.m]: I don't see how your reply is in any way related to my comment. You probably wanted to explain that this code is not yours, you just moved it into a separate file.

Comment (nongio) [reply Source/wayland/WaylandServer+Cursor.m]: yes, the comment is more related to the general feedback that the PR is very large to be reviewed. Anyway yes, the Cursor code has only been moved in a new file

**Comment by gcasa** [comment]
Merging this on behalf of @nongio since he doesn't have merge permission.  I spoke with him via email.

**Pull request merged**
